### PR TITLE
Clarify normative statement related to timezone offsets.

### PR DESCRIPTION
**Pull request by msporny**
This PR attempts to address issue #1343 by clarifying the normative statement related to timezone offsets.

/cc @jyasskin

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/vc-data-model/pull/1344.html" title="Last updated on Nov 12, 2023, 10:42 PM UTC (3cb0665)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/vc-data-model/1344/998a3dc...3cb0665.html" title="Last updated on Nov 12, 2023, 10:42 PM UTC (3cb0665)">Diff</a>

### CODE REVIEW

**Review by jyasskin — APPROVED** [review]
Thanks, this is clearer.

File: index.html
Comment (aphillips) [inline index.html:1924]: This is fine, albeit a lot of detail for your spec. 

This doesn't say what happens when a given VC credential ignores it (is the credential invalid??) One work around for this would be to specify that time values must include an offset (with UTC preferred) and separately specify that values without an offset will be interpreted as UTC. 

Also, note that temporal serializations are coming, which include true time zone IDs (offsets have many problems). The I-D isn't an RFC yet, so you can't reference it here, but be aware of it for revisions in the future. The WG there is called [SEDATE](https://datatracker.ietf.org/doc/draft-ietf-sedate-datetime-extended/)

**Review by iherman — APPROVED** [review]

File: index.html
Comment (msporny) [reply index.html:1924]: > This is fine, albeit a lot of detail for your spec.

Yes. :P

When people raise issues, we try to be crystal clear in the changes, and that has led to a verbose specification (that some people are cranky about)... but if we remove the detail, we end up not addressing the concerned party's needs/concerns. Such is the specification editing process. :)

> This doesn't say what happens when a given VC credential ignores it (is the credential invalid??) One work around for this would be to specify that time values must include an offset (with UTC preferred) and separately specify that values without an offset will be interpreted as UTC.

Good idea, that covers all of our bases AFAICT. I'll implement it in an upcoming modification to this PR.

It would allow us to weaken the existing MUST to a SHOULD, but still ensure a single "correct" way to interpret all time values. I'm on the fence wrt. weakening the existing MUST to a SHOULD -- thoughts, @jyasskin and @aphillips? I could go either way.

> Also, note that temporal serializations are coming, which include true time zone IDs (offsets have many problems). The I-D isn't an RFC yet, so you can't reference it here, but be aware of it for revisions in the future. The WG there is called [SEDATE](https://datatracker.ietf.org/doc/draft-ietf-sedate-datetime-extended/)

Interesting... if we were to switch to IXDTF syntax in the future, we'd have to move away from XML Schema dateTimeStamp (and the ref to that spec) and move back to RFC3339 (which we had to move away from for reasons I can't remember right now). That is definitely a VC v3.0 thing, so let's let things settle in SEDATE-land before we make any changes. I don't forsee any spec issues if we had to switch given that dateTimeStamps are a subset of IXDTF syntax. The biggest issue is deployed infrastructure choking on IXDTF datetimes... we'd have to make sure IXDTF parsing is in a large variety of datetime libraries before shipping that feature.

Comment (aphillips) [reply index.html:1924]: > It would allow us to weaken the existing MUST to a SHOULD, but still ensure a single "correct" way to interpret all time values. I'm on the fence wrt. weakening the existing MUST to a SHOULD -- thoughts, @jyasskin and @aphillips? I could go either way.

I'd be reluctant to reduce MUST to SHOULD as well. Perhaps something like:

```suggestion
All time values MUST be specified as dateTimeStamp, either in UTC (denoted by `Z`) 
or with a time zone offset relative to Universal Coordinated Time. 
Time values incorrectly serialized without an offset MUST be interpreted as UTC.
```

(above suggestion is not meant to be merged, as it needs references and such)

> Interesting... if we were to switch to IXDTF syntax in the future, we'd have to move away from XML Schema dateTimeStamp (and the ref to that spec) and move back to RFC3339 (which we had to move away from for reasons I can't remember right now). That is definitely a VC v3.0 thing, so let's let things settle in SEDATE-land before we make any changes. I don't forsee any spec issues if we had to switch given that dateTimeStamps are a subset of IXDTF syntax. The biggest issue is deployed infrastructure choking on IXDTF datetimes... we'd have to make sure IXDTF parsing is in a large variety of datetime libraries before shipping that feature.

I suspect Temporal will be as popular in JS as it has been in other programming environments. This will likely lead to various serializations getting updated too. RFC3339 is just ahead of the curve here. You have the right idea: wait for the dust to settle.

**Review by TallTed — APPROVED** [review]
LGTM after tweaks as requested by @aphillips

File: index.html
Comment (msporny) [reply index.html]: Applied your changes in 91339e127d533724a276663a805ad7db882043d7.

**Review by longpd — APPROVED** [review]

**Review by OR13 — APPROVED** [review]

**Review by aphillips — APPROVED** [review]
LGTM! Thanks!

**Review by dlongley — APPROVED** [review]

**Review by decentralgabe — APPROVED** [review]

**Comment by jandrieu** [comment]
Seems overly broad to *all* time values. Would infect custom values.

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/2017/vc/WG/Meetings/Minutes/2023-11-14-vcwg-special#section1-10) on 2023-11-14

- no resolutions were taken

<details><summary><i>View the transcript</i></summary>

#### 1.10. Clarify normative statement related to timezone offsets. (pr vc-data-model#1344)

_See github pull request [vc-data-model#1344](https://github.com/w3c/vc-data-model/pull/1344)._




**Brent Zundel:** this PR is broadly approved with no unresolved change requests.  
… we are going to merge in a couple of days.  

**Manu Sporny:** heads up to the group, this is also significant. When you express a time value in any VC, you should express it as a date timestamp formatted value.  
… anyone verifying, if they do not see a relative UTC offset, has to interpret it as UTC.  
… this means you cannot express (should not express) a date that does not have a timezone offset at the end of it for any time value in a VC.  

**Joe Andrieu:** also commented in github - would like to see this scoped to just the properties we are defining in the VCDM. Don't want to see this on custom properties, or.  

**Manu Sporny:** that is the intention. we are trying to increase interop.  

**Brent Zundel:** let's continue the conversation on the PR.  
… please be aware of and review other PRs that we didn't get a chance to review today.  

---

</details>

**Comment by msporny** [comment]
> @msporny You seemed to say on the call today that it is the intention of this PR to apply to included vocabularies used for claims in VCs. So that we are explicitly discouraging extending a VC with a formally defined @context with its own definitions for various temporal properties. [Ivan thought you misheard me, so please clarify if I misheard you.]

I think you misheard me, but maybe not... here's what I was saying:

I want people to pay special attention to this PR because we're providing guidance on how people should express time values. We are saying two things:

1. If you're generating conforming documents, you SHOULD use dateTimeStamp to express time values. 
2. If you are consuming conforming documents, and someone has expressed a time in a way that doesn't specify any timezone/timeoffset information, you MUST presume the timezone/timeoffset is UTC.

We are not saying that someone can't extend the data model and have properties for `myTimeIsBetterThanYours` or `asnValidFrom`. We're just saying that when they do that, please use `dateTimeStamp` to express the value, and regardless of whether you do that or not, that time value, when read, and if it doesn't include a timezone, will be interpreted as UTC.

So, we're trying to strike a balance here to 1) not be overbearing in how extensions express time values (but provide strong guidance), and 2) be very strict in giving people fair warning that if they don't include a timezone/offset, that their times will be interpreted as UTC.

> I believe this would be the first time we are suggesting restricting extended vocabularies beyond locking down specific properties of the VCDM.

We already tell people where to put properties that express things on the VC vs. the credential subject, and where to put fields that are associated with extension points like `credentialStatus` -- those objects MUST contain (`type`, at a minimum). So, I do think we specify what extensions MUST do in some cases.

> The result would be to bias the ecosystem against vocabularies that have a reason to represent time differently.

I mean, we still allow it... what's the compelling use case for representing time differently? Remember that we've had decades of standards on trying to NOT represent time differently (ISO8601, RFC3339, XML Datetime) -- they all (more or less) follow the same format with some fairly esoteric differences.

> That would be an untenable position, given our commitment to allow anyone to say anything.

We still allow anyone to say anything. I don't know how this PR would prevent that?

> This restriction would say that although you are conceptually ALLOWED to say anything, you SHOULDN'T. Instead you SHOULD only use vocabularies that meet certain rules.

The statement above seems much stronger than what the PR is doing... don't know if any of the explanatory text above helped?

> This is not just a bad idea, IMO, it violates the core premise and promise of VCs. As long as the extensions to the VCDM do not override required values, they should be encouraged as first class uses of the technology instead of biasing the ecosystem against such extendibility.

I don't think we're biasing against extensibility -- we're just saying, if you express time values, you SHOULD do them in this way. We're doing that to increase interop, because this specific thing (not including timezone information) has shown up in deployed systems and it's just a matter of time before it starts biting the ecosystem.

Hope the above helps, if not, and your concern is a blocking concern, let me know sooner than later.

**Comment by dlongley** [comment]
> We are not saying that someone can't extend the data model and have properties for `myTimeIsBetterThanYours` or `asnValidFrom`. We're just saying that when they do that, please use `dateTimeStamp` to express the value, and regardless of whether you do that or not, that time value, when read, and if it doesn't include a timezone, will be interpreted as UTC.

I would tweak this by saying: "...when read, and if it doesn't include a timezone, and no custom advice is given for `myTimeIsBetterThanYours` for how to handle that, it will be interpreted as UTC."

**Comment by TallTed** [comment]
I'm with @dlongley, with this addition --
`s/when read, and if it doesn't/when read, if it doesn't/`
(i.e., drop the misplaced `and`)

**Comment by msporny** [comment]
I'm waiting to hear back from @jandrieu to ensure he's not objecting before merging this PR.

**Review by jandrieu — APPROVED** [review]
I'm good with this as a SHOULD.

**Comment by msporny** [comment]
Normative, multiple reviews, changes requested and made, no objections, merging.

**Pull request merged**
